spacy/tokens/doc.pyx

```diff
@@ -1451,6 +1451,70 @@ cdef class Doc:
                 remove_label_if_necessary(attributes[i])
                 retokenizer.merge(span, attributes[i])
 
+    @staticmethod
+    def from_json(vocab, doc_json):
+        """Convert a JSON document generated by Doc.to_json() to a Doc.
+
+        vocab (Vocab): A storage container for lexical types.
+        doc_json (dict): The Doc data in JSON format.
+        RETURNS (Doc): A doc instance corresponding to the specified JSON representation.
+        """
+
+        words = []
+        token_annotations = {
+            # For each annotation type: store (1) annotation type required to include it, (2) list of annotations for
+            # this type.
+            annotation_type: {"req": annotation_type if annotation_type != "head" else "dep", "values": None}
+            for annotation_type in ("tag", "pos", "morph", "lemma", "dep", "head")
+        }
+
+        # Gather token-level properties.
+        for token in doc_json["tokens"]:
+            words.append(doc_json["text"][token["start"]: token["end"]])
+            for annotation_type in token_annotations:
+                # Assuming all tokens have exactly the same set of attributes.
+                if token_annotations[annotation_type]["req"] in token:
+                    if token_annotations[annotation_type]["values"] is None:
+                        token_annotations[annotation_type]["values"] = []
+                    token_annotations[annotation_type]["values"].append(token[annotation_type])
+
+        # Create Doc instance.
+        doc = Doc(
+            vocab,
+            words=words,
+            tags=token_annotations["tag"]["values"],
+            pos=token_annotations["pos"]["values"],
+            morphs=token_annotations["morph"]["values"],
+            lemmas=token_annotations["lemma"]["values"],
+            heads=token_annotations["head"]["values"],
+            deps=token_annotations["dep"]["values"],
+        )
+
+        # Complement other document-level properties (cats, spans, ents).
+        doc.cats = doc_json.get("cats", doc.cats)
+
+        if "spans" in doc_json:
+            for span_group in doc_json["spans"]:
+                spans = []
+                for json_span in doc_json["spans"][span_group]:
+                    span = doc.char_span(json_span["start"], json_span["end"], json_span["label"], json_span["kb_id"])
```

Comment:
I would add an `assert span is not None` here, to guard against internal programming errors.

Reply:
Included in next commit.